### PR TITLE
The big PR

### PR DESCRIPTION
**Pull request by aenario**
@jsilvestre reviewed up to commit 42d5846

Need more comments.

### CODE REVIEW

File: client/app/components/account_config.coffee [hunk 1]
```diff
@@ -1,12 +1,30 @@
...
+
+REQUIRED_FIELDS_NEW = [
+    'label', 'name', 'login', 'password', 'imapServer', 'imapPort',
+    'smtpServer', 'smtpPort', 'smtpMethod'
+]
+
+REQUIRED_FIELDS_EDIT = REQUIRED_FIELDS_NEW
+# .concat [
+#     'draftMailbox', 'sentMailbox', 'trashMailbox'
+# ]
```
Comment (nono) [inline client/app/components/account_config.coffee]: @aenario is it normal this commented code?

File: client/app/components/account_config.coffee [hunk 2]
```diff
@@ -15,390 +33,144 @@ module.exports = React.createClass
...
+        errors = @getLocalValidationErrors @state.editedAccount
 
-        {accountValue, valid, errors} = @validateForm()
+        id = @state.editedAccount?.get('id')
+        accountValue = @state.editedAccount.toJS()
 
-        if Object.keys(errors).length > 0
+        if errors.length > 0
             LayoutActions.alertError t 'account errors'
+            @setState {submitted: true, errors: errors}
```
Comment (nono) [inline client/app/components/account_config.coffee]: @aenario why the `{` `}` ?

File: client/app/components/account_config_mailboxes.coffee [hunk 3]
```diff
@@ -4,150 +4,68 @@ classer = React.addons.classSet
...
-        return state
+        newMailboxName: ''
+        newMailboxParent: null
 
+    makeLinkState: (field) ->
+        currentValue = @props.editedAccount.get(field)
+        cachedTransform @, '__cacheLS', currentValue, =>
+            value: currentValue
+            requestChange: (value) =>
+                (changes = {})[field] = value
```
Comment (nono) [inline client/app/components/account_config_mailboxes.coffee]: @aenario it's hard to read this line. It's the same than `changes = "#{field}": value`?

File: client/app/components/account_config_mailboxes.coffee [hunk 4]
```diff
@@ -158,57 +76,98 @@ module.exports = AccountConfigMailboxes = React.createClass
...
+            div className: 'col-sm-3',
+                MailboxPicker
+                    allowUndefined: false
+                    valueLink: @makeLinkState property
+                    mailboxes: @props.editedAccount.get('mailboxes')
 
     onMailboxChange: (mailbox, box) ->
-        @props[box].requestChange mailbox, =>
+        changes = {}
+        changes[box] = mailbox
```
Comment (nono) [inline client/app/components/account_config_mailboxes.coffee]: @aenario `changes` seems to be not used. Am I missing something?

File: client/app/components/account_config_mailboxes.coffee [hunk 5]
```diff
@@ -158,57 +76,98 @@ module.exports = AccountConfigMailboxes = React.createClass
...
+                    allowUndefined: true
+                    valueLink: @linkState 'newMailboxParent'
+                    mailboxes: @props.editedAccount.get('mailboxes')
+
+    renderMailboxChoice: (which) ->
+        property = "#{which}Mailbox"
+        labelText = t "account #{which} mailbox"
+
+        className = "form-group #{which}"
+        className += 'has-error' unless @props.editedAccount.get property
```
Comment (nono) [inline client/app/components/account_config_mailboxes.coffee]: @aenario is it normal to have `#{which}` and `has-error` with no space to separate them?

File: client/app/components/account_config_main.coffee [hunk 6]
```diff
@@ -27,176 +37,120 @@ module.exports = AccountConfigMain = React.createClass
...
+        options.name ?= field
+        options.key = "account-config-field-#{field}"
+        options.valueLink ?= @makeLinkState field
+        options.error ?= @props.errors?.get field
+        return AccountInput options
 
     render: ->
-        buttonLabel = @buildButtonLabel()
-
+        console.log 'ACMRENDER'
```
Comment (nono) [inline client/app/components/account_config_main.coffee]: @aenario  `console.log`

File: client/app/components/account_config_main.coffee [hunk 7]
```diff
@@ -27,176 +37,120 @@ module.exports = AccountConfigMain = React.createClass
...
+                @buildInput 'label'
+                @buildInput 'name'
+                @buildInput 'login' #, type: 'email'
 
-            AccountInput
-                name: 'label'
-                value: @linkState('label').value
-                errors: @state.errors
-                onBlur: @props.onBlur
+                if not isOauth
```
Comment (nono) [inline client/app/components/account_config_main.coffee]: @aenario s/if not/unless/ ?

File: client/app/components/account_config_main.coffee [hunk 8]
```diff
@@ -268,214 +179,85 @@ module.exports = AccountConfigMain = React.createClass
...
-
-                if type is 'ssl' and target.checked
-                    @setState smtpPort: 465
-
-                else if type is 'tls' and target.checked
-                    @setState smtpPort: 587
-
+    doDiscovery: (domain) ->
+        if domain? and domain.length > 3 and domain isnt @_lastDiscovered
+            console.log domain, @discoverTimeout
```
Comment (nono) [inline client/app/components/account_config_main.coffee]: @aenario `console.log`

File: client/app/components/account_config_main.coffee [hunk 9]
```diff
@@ -268,214 +179,85 @@ module.exports = AccountConfigMain = React.createClass
...
+            if @discoverTimeout
+                @nextDiscover = domain
             else
+                @discoverTimeout = setTimeout (=> @doDiscoveryNow domain), 2000
```
Comment (nono) [inline client/app/components/account_config_main.coffee]: @aenario maybe using something like [`throttle` from lodash](https://lodash.com/docs#throttle) can help to simplify your code

File: client/app/components/conversation.coffee [hunk 10]
```diff
@@ -118,29 +133,29 @@ module.exports = React.createClass
 
 
     render: ->
-        if not @props.conversation
+        console.log "RENDER", uniq
```
Comment (nono) [inline client/app/components/conversation.coffee]: @aenario `console.log`

File: client/app/mixins/should_update_mixin.coffee [hunk 11]
```diff
@@ -0,0 +1,28 @@
+
+shallowEqual = (A, B, log) ->
```
Comment (nono) [inline client/app/mixins/should_update_mixin.coffee]: @aenario is the `log` still used?

Comment (aenario) [inline client/app/mixins/should_update_mixin.coffee]: The log is used in the debugging ShouldComponentUpdate.Logging mixin (below)

File: client/app/components/account_config_main.coffee (hunk 9 above)
Comment (aenario) [inline client/app/components/account_config_main.coffee]: I tried a bit, but had trouble interfacing it with React autobinding. This code will need moving to a proper actions/store anyway. I will try again to use _.throttle or _.debounce at this point.

**Comment by poupotte** [comment]
@aenario , can you update your ticket (https://huboard.com/cozy/cozy-emails#/) ? (Move tickets solved in this PR in 'To Merge' )

**Comment by frankrousseau** [comment]
Can we merge it?

**Pull request merged**